### PR TITLE
Piksifaulthandler

### PR DESCRIPTION
**Pull request by 18kudlugim**
Piksi Fault Handler checks three conditions before moving mission state to standby.

After shifting the last_rtkfix_ccno field from PiksiControlTask.cpp to PiksiFaultHandler.cpp, it's failing to pass some unit tests [WIP]. Regarding the PTest, checking the third condition after having checked the second condition results in the mission state being prematurely set to standby. However, since the unit tests aren't passing, this could have something to do with the PiksiFaultHandler implementation.

### CODE REVIEW

File: test/test_fsw_mission_manager/test_fixture.cpp [hunk 1]
```diff
@@ -38,7 +37,7 @@ TestFixture::TestFixture(mission_state_t initial_state, unsigned int bootcount)
     adcs_dcdc_fp = registry.create_writable_field<bool>("dcdc.ADCSMotor_cmd");
 
     piksi_state_fp = registry.create_readable_field<unsigned char>("piksi.state");
-    last_rtkfix_ccno_fp = registry.create_internal_field<unsigned int>("piksi.last_rtkfix_ccno");
+    // last_rtkfix_ccno_fp = registry.find_internal_field_t<unsigned int>("piksi.last_rtkfix_ccno");
```
Comment (shihaocao) [inline test/test_fsw_mission_manager/test_fixture.cpp:40]: i think mission manager still needs this field, except now its coming from the fault handler

Comment (shihaocao) [reply test/test_fsw_mission_manager/test_fixture.cpp:40]: and now its readable

File: ptest/configs/hitl_singlesat.json [hunk 2]
```diff
@@ -19,8 +19,13 @@
             "connect": false,
             "send_queue_duration": 0,
             "send_lockout_duration": 0,
+<<<<<<< Updated upstream
             "check_uplink_queue_enable": false,
             "http_port": 8001
+=======
+            "http_port": 8001,
+            "check_uplink_queue_enable": false
+>>>>>>> Stashed changes
```
Comment (kylekrol) [inline ptest/configs/hitl_singlesat.json:28]: We committed a merge conflict here.

File: src/fsw/FCCode/Estimators/RelativeOrbitEstimator.cpp [hunk 3]
```diff
@@ -14,24 +14,25 @@
...
 static constexpr lin::Matrixd<6, 6> sqrtQ = lin::diag(sqrtQ_diag);
 
 // Estimator sensor noise
 TRACKED_CONSTANT_SC(double, RelOrbit_sqrtR_r, 1.0e-1);
 static constexpr lin::Vector3d sqrtR_diag =
-        RelOrbit_sqrtR_r * lin::ones<lin::Vector3d>();
+    RelOrbit_sqrtR_r * lin::ones<lin::Vector3d>();
 static constexpr lin::Matrix3x3d sqrtR = lin::diag(sqrtR_diag);
 
 static inline auto get_gps_zero()
```
Comment (kylekrol) [inline src/fsw/FCCode/Estimators/RelativeOrbitEstimator.cpp:38]: Can we undo the changes here that are purely formatting? It's just going to clutter the git blame and it'll be more difficult to track down whose actually responsible for what code in the future.

File: ptest/configs/hitl_singlesat.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
         {
             "name": "FlightController",
             "run_mode": "teensy",
-            "port": "/dev/ttyACM0",
+            "port": "/dev/cu.usbmodem73392101",
```
Comment (kylekrol) [inline ptest/configs/hitl_singlesat.json:7]: Let's also not commit this change here as it's different for everyone.

File: src/fsw/FCCode/PiksiControlTask.cpp [hunk 5]
```diff
@@ -5,39 +5,39 @@
...
-    microdelta_f("piksi.microdelta", Serializer<unsigned int>()),
-    last_rtkfix_ccno_f("piksi.last_rtkfix_ccno")
-    {
-        add_readable_field(pos_f);
-        add_readable_field(vel_f);
-        add_readable_field(baseline_pos_f);
-        add_readable_field(current_state_f);
-        add_readable_field(fix_error_count_f);
-        add_readable_field(time_f);
-        add_readable_field(microdelta_f);
```
Comment (kylekrol) [inline src/fsw/FCCode/PiksiControlTask.cpp:27]: Same with this file, we shouldn't do a bunch of formatting stuff without make a substantive change in the file.

File: src/fsw/FCCode/PiksiControlTask.hpp [hunk 6]
```diff
@@ -44,7 +44,7 @@ class PiksiControlTask : public TimedControlTask<void>
     ReadableStateField<unsigned int> microdelta_f;
 
     // Control cycle of last good Piksi reading
-    InternalStateField<unsigned int> last_rtkfix_ccno_f;
+    // InternalStateField<unsigned int> last_rtkfix_ccno_f;
```
Comment (kylekrol) [inline src/fsw/FCCode/PiksiControlTask.hpp:47]: If we don't need it we can just delete it?

File: ptest/configs/hitl_singlesat.json (hunk 4 above)
Comment (18kudlugim) [reply ptest/configs/hitl_singlesat.json:7]: whoops didn't realize I commited this

Comment (shihaocao) [reply ptest/configs/hitl_singlesat.json:7]: bump to remove this

File: src/fsw/FCCode/PiksiControlTask.cpp (hunk 5 above)
Comment (shihaocao) [reply src/fsw/FCCode/PiksiControlTask.cpp:27]: bump to checkout this file from master

File: src/fsw/FCCode/PiksiFaultHandler.hpp [hunk 7]
```diff
@@ -4,28 +4,33 @@
...
-class PiksiFaultHandler : public FaultHandlerMachine {
-  public:
-    //! Default maximum wait times
-    TRACKED_CONSTANT_SC(unsigned int, default_no_cdgps_max_wait, PAN::one_day_ccno);
-    TRACKED_CONSTANT_SC(unsigned int, default_cdgps_delay_max_wait, PAN::one_day_ccno/8);
+#include <common/GPSTime.hpp>
+#include <common/constant_tracker.hpp>
 
-    /**
+#include "TimedControlTask.hpp"
```
Comment (shihaocao) [inline src/fsw/FCCode/PiksiFaultHandler.hpp:10]: bump to undo formatting overall on this file

File: src/fsw/FCCode/PiksiFaultHandler.hpp [hunk 8]
```diff
@@ -4,28 +4,33 @@
 #include "FaultHandlerMachine.hpp"
 #include "piksi_mode_t.enum"
 
-class PiksiFaultHandler : public FaultHandlerMachine {
-  public:
-    //! Default maximum wait times
-    TRACKED_CONSTANT_SC(unsigned int, default_no_cdgps_max_wait, PAN::one_day_ccno);
-    TRACKED_CONSTANT_SC(unsigned int, default_cdgps_delay_max_wait, PAN::one_day_ccno/8);
+#include <common/GPSTime.hpp>
+#include <common/constant_tracker.hpp>
```
Comment (shihaocao) [inline src/fsw/FCCode/PiksiFaultHandler.hpp:8]: we should keep these only if they are critical for build

File: src/fsw/FCCode/PiksiFaultHandler.cpp [hunk 9]
```diff
@@ -4,63 +4,79 @@
...
 
     piksi_mode_t piksi_state = static_cast<piksi_mode_t>(piksi_state_fp->get());
     mission_state_t mission_state = static_cast<mission_state_t>(mission_state_fp->get());
 
-    if (piksi_state == piksi_mode_t::dead) {
-        return fault_response_t::standby;
+    // Recommend moving to safehold if piksi is dead in a non-standby state
+    if (mission_state != mission_state_t::startup && piksi_state == piksi_mode_t::dead)
+    {
+        return fault_response_t::safehold;
```
Comment (shihaocao) [inline src/fsw/FCCode/PiksiFaultHandler.cpp:45]: @18kudlugim @madeleinemc This fault response will break pending both PR's being merged in. As soon as we flip into detumble, the piksi would still be dead.

Maybe this is the time we finally remove the piksi dead state, and replace it with a fault that's signaled whenever the piksi_state is in an error state. And then only incrementing signaling whenever we're not in standby

File: src/fsw/FCCode/PiksiFaultHandler.cpp [hunk 10]
```diff
@@ -4,63 +4,79 @@
...
     }
 
-    if (mission_state == mission_state_t::follower_close_approach || 
-            mission_state == mission_state_t::leader_close_approach) {
+    if (piksi_state == piksi_mode_t::fixed_rtk)
+    {
+        last_rtkfix_ccno_f.set(count_cc);
+    }
```
Comment (shihaocao) [inline src/fsw/FCCode/PiksiFaultHandler.cpp:51]: this is good

File: src/fsw/FCCode/PiksiFaultHandler.cpp [hunk 11]
```diff
@@ -4,63 +4,79 @@
...
-PiksiFaultHandler::PiksiFaultHandler(StateFieldRegistry& r) 
-    : FaultHandlerMachine(r), 
-    no_cdgps_max_wait_f("piksi_fh.no_cdpgs_max_wait", Serializer<unsigned int>(PAN::one_day_ccno)),
-    cdgps_delay_max_wait_f("piksi_fh.cdpgs_delay_max_wait", Serializer<unsigned int>(PAN::one_day_ccno)),
-    fault_handler_enabled_f("piksi_fh.enabled", Serializer<bool>())
-    {
-        add_writable_field(no_cdgps_max_wait_f);
-        add_writable_field(cdgps_delay_max_wait_f);
-        add_writable_field(fault_handler_enabled_f);
+PiksiFaultHandler::PiksiFaultHandler(StateFieldRegistry &r)
```
Comment (shihaocao) [inline src/fsw/FCCode/PiksiFaultHandler.cpp:9]: bump to undo formatting changes

**Review by shihaocao — CHANGES_REQUESTED** [review]
Biggest thing is the fact that we're going to have an upcoming problem with the PiksiFaultHandler potentially interacting badly with the PiksiDontStart code package change. @madeleinemc @18kudlugim could you both update each other on this?

File: src/fsw/FCCode/PiksiControlTask.cpp (hunk 5 above)
Comment (18kudlugim) [reply src/fsw/FCCode/PiksiControlTask.cpp:27]: Ahh it autoformatted when I saved :')

Comment (18kudlugim) [reply src/fsw/FCCode/PiksiControlTask.cpp:27]: I am removing all instances of last_rtkfix_ccno_f in the file. Does that count as a substantive change?

File: src/fsw/FCCode/PiksiFaultHandler.cpp (hunk 11 above)
Comment (18kudlugim) [reply src/fsw/FCCode/PiksiFaultHandler.cpp:9]: Are these just formatting changes? I've added the last_rtkfix_ccno_f state field in the constructor and all, that's why the lines have shifted down

Comment (shihaocao) [reply src/fsw/FCCode/PiksiFaultHandler.cpp:9]: Yeah sorry like, indeed your line is import and should remain, but we shouldn't reformat all the other lines. in this file that are not related to your change.

**Review by shihaocao — APPROVED** [review]
Sick

**Pull request merged**